compiler/rustc_const_eval/src/const_eval/machine.rs

```diff
@@ -162,12 +162,24 @@ pub(crate) type CompileTimeEvalContext<'mir, 'tcx> =
 #[derive(Debug, PartialEq, Eq, Copy, Clone)]
 pub enum MemoryKind {
     Heap,
+    /// The base allocation of a static.
+    /// When this is read from, we need to error, because otherwise we could
+    /// invent values from nothing (e.g. `static FOO: Thing = FOO;`). We need
+    /// to do this for all statics, to prevent static items from inventing values
+    /// by reading ZSTs from each other:
+    /// ```rust,compile_fail
+    /// struct Thing;
+    /// static FOO: Thing = BAR;
+    /// static BAR: Thing = FOO;
+    /// ```
+    StaticRoot(DefId),
```

Comment:
Do we need the DefId? Can't we just call this `CurrentStaticRoot` and make `before_memory_read` error when encountering a read from such an allocation?

And in fact, do we have to do this via the MemoryKind? Can't this be done by storing the `AllocId` of the "blocked" allocation in the CTFE machine and just preventing all reads from that? That would require changing way fewer things, I think, and might also avoid all the trouble with changing the MemoryKind later?

Reply:
got rid of it